doc/hardware/clock_mgmt/index.rst

```diff
@@ -0,0 +1,771 @@
+  needed by this clock node (IE a C structure)
+
+* ``Z_CLOCK_MGMT_VND_CLOCK_NODE_DATA_GET``: gets a reference to any static
+  structure defined by the ``DATA_DEFINE`` macro. This is used to initialize the
+  ``void *`` passed to :c:func:`clock_configure`, so for many clock nodes this
+  macro can simply expand to an integer value (which may be used for a register
+  setting)
+
+As an example, for the following devicetree:
+
+.. code-block:: devicetree
+
+    clock_source: clock-source {
+        compatible = "fixed-clock";
+        clock-frequency = <10000000>;
+        #clock-cells = <0>;
+
+        clock_div: clock-div@50000000 {
+            compatible = "vnd,clock-div";
+            #clock-cells = <1>;
+            reg = <0x5000000>;
+
+            clock_output: clock-output {
+                compatible = "clock-output";
+                #clock-cells = <0>;
+            };
+        };
+    };
+
+    ....
+
+    &clock_output {
+        clock_output_state_default: clock-output-state-default {
+            compatible = "clock-state";
+            clocks = <&clock_div 1>;
+            clock-frequency = <DT_FREQ_M(10)>
+        };
+    }
+
+    ....
+
+    periph0: periph@0 {
+        /* Clock outputs */
+        clock-outputs= <&clock_output>;
+        clock-output-names = "default";
+        /* Default clock state */
+        clock-state-0 = <&clock_output_state_default>;
+        clock-state-names = "default";
+    };
+
+The clock subsystem would expect the following macros be defined:
+
+* ``Z_CLOCK_MGMT_VND_CLOCK_DIV_DATA_DEFINE``
+* ``Z_CLOCK_MGMT_VND_CLOCK_DIV_DATA_GET``
+
+These macros should be defined within a header file. The header file can then
+be added to the list of clock management driver headers to include using the
+CMake function ``add_clock_mgmt_header`` or ``add_clock_mgmt_header_ifdef``.
+
+Output Clock Nodes
+------------------
+
+Clock trees should define output clock nodes as leaf nodes within their
+devicetree. These nodes must have the compatible :dtcompatible:`clock-output`,
+and are the nodes which clock consumers will reference. The clock management
+framework will handle defining clock drivers for each of these nodes.
+
+Common Clock Drivers
+--------------------
+
+For some common clock nodes, generic drivers already exist to simplify vendor
+implementations. For a list, see the table below:
+
+
+.. table:: Common Clock Drivers
+    :align: center
+
+    +-------------------------------------+--------------------------------------------+
+    | DT compatible                       | Use Case                                   |
+    +-------------------------------------+--------------------------------------------+
+    | :dtcompatible:`fixed-clock`         | Fixed clock sources that cannot be gated   |
+    +-------------------------------------+--------------------------------------------+
+    | :dtcompatible:`clock-source`        | Gateable clock sources with a fixed rate   |
```

Comment:
This naming is a bit confusing. `Fixed-clock' is usually used for fixed-rate clocks sources. 
With`clock-source` is used as opposition to `fixed-clock`, one could quickly think about "non fixed rate clock".

What about `fixed-clock-always-on` and `fixed-clock`?
Or even an optional `always-on` property in `fixed-clock` binding?

Reply:
I think we have to keep `fixed-clock` as is, because Linux has already set this binding in stone: https://github.com/torvalds/linux/blob/87d6aab2389e5ce0197d8257d5f8ee965a67c4cd/Documentation/devicetree/bindings/clock/fixed-clock.yaml#L1

For `clock-source`, we could split this into a `fixed-clock` and generic `clock-gate` binding. I kept this as one driver because it will likely result in less flash consumption- the more granular we are about splitting clock configurations into different nodes, the more flash we use (generally, as we allocate more clock structures)

Reply:
As an aside, many of the clock bindings won't be inherited from Linux, because Linux does not declare each clock node in devicetree the way this PR does.